src/main/java/io/cdap/plugin/gcp/bigquery/relational/BigQueryRelation.java

```diff
@@ -1,117 +1,409 @@
+    Map<String, Expression> selectedColumns = getSelectedColumns(columns);
+    // Build new transform expression and return new instance.
+    Supplier<String> supplier =
+      () -> buildNestedSelect(selectedColumns, getTransformExpression(), newDatasetName, null);
+    return new BigQueryRelation(newDatasetName, columns, this, supplier);
   }
 
   @Override
   public Relation setColumn(String column, Expression value) {
-    Map<String, Expression> newColumns = new HashMap<>(columns);
-    newColumns.put(column, value);
+    // check if expression is supported and valid
+    if (!supportsExpression(value)) {
+      return getInvalidRelation("Unsupported or invalid expression type.");
+    }
 
-    return new BigQueryRelation(datasetName, newColumns, filter, this);
+    Map<String, Expression> selectedColumns = getSelectedColumns(columns);
+    selectedColumns.put(column, value);
+
+    // Build new transform expression and return new instance.
+    Supplier<String> supplier =
+      () -> buildNestedSelect(selectedColumns, getTransformExpression(), datasetName, null);
+    return new BigQueryRelation(datasetName, selectedColumns.keySet(), this, supplier);
   }
 
   @Override
   public Relation dropColumn(String column) {
-    Map<String, Expression> newColumns = new HashMap<>(columns);
-    newColumns.remove(column);
+    // check if all expressions are supported and valid
+    if (!columns.contains(column)) {
+      return getInvalidRelation("Trying to remove non existing column in Relation: " + column);
+    }
+
+    // Remove column
+    Map<String, Expression> selectedColumns = getSelectedColumns(columns);
+    selectedColumns.remove(column);
 
-    return new BigQueryRelation(datasetName, newColumns, filter, this);
+    // Build new transform expression and return new instance.
+    Supplier<String> supplier =
+      () -> buildNestedSelect(selectedColumns, getTransformExpression(), datasetName, null);
+    return new BigQueryRelation(datasetName, selectedColumns.keySet(), this, supplier);
   }
 
   @Override
   public Relation select(Map<String, Expression> columns) {
-    return new BigQueryRelation(datasetName, columns, filter, this);
+    // check if all expressions are supported and valid
+    if (!supportsExpressions(columns.values())) {
+      return getInvalidRelation("Unsupported or invalid expression type.");
+    }
+
+    // Build new transform expression and return new instance.
+    Supplier<String> supplier =
+      () -> buildNestedSelect(columns, getTransformExpression(), datasetName, null);
+    return new BigQueryRelation(datasetName, columns.keySet(), this, supplier);
   }
 
   @Override
   public Relation filter(Expression filter) {
-    return new BigQueryRelation(datasetName, columns, filter, this);
+    // check if expression is supported and valid
+    if (!supportsExpression(filter)) {
+      return getInvalidRelation("Unsupported or invalid expression type.");
```

Comment:
It would be great to propagate expression validity message here (and in other calls) if expression is not valid. Otherwise it would be hard to debug.

Reply:
Can be done in follow up PR though to get this one out